src/core/tensor/tensor.cc

```diff
@@ -1050,6 +1107,46 @@ void Mult(const SType alpha, const Tensor &A, const Tensor &B, const SType beta,
   }
 }
 
+  // ************************
+  // Some traversal operations (works on const declarations without modifying tensor variables)
+  // ************************
+
+  //generate shape multipliers
+  std::vector<int> generate_shape_multipliers(std::vector<int> y_shape) {
+      std::reverse(y_shape.begin(), y_shape.end());
+      std::vector<int> shape_multipliers = {};
+      int cumulative_product = 1;
+
+      shape_multipliers.push_back(1);
+      for (int n=0; n<(y_shape.size()-1); ++n) {
+          cumulative_product = cumulative_product*y_shape[n];
+          shape_multipliers.push_back(cumulative_product);
+      }
+      std::reverse(shape_multipliers.begin(), shape_multipliers.end());
+      return shape_multipliers;
+  };
+
+  int determine_order(size_t counter) const {
+      for (int n=0; n<(shape_multiplier_.size()-1); ++n) {
+          if((counter%shape_multiplier_[n])==0){
+              return ((shape_multiplier_.size()) - 1 - n);
+          }
+      }
+      return 0;
+  };
+
+  void update_base_index(std::vector<int>& traversal_info) const {
+      for (int n=0; n<(traversal_info[3]+1); ++n) {
+          traversal_info[n] = traversal_info[2];
+      }
+  };
+
+  void traverse_next(std::vector<int>& traversal_info, size_t&& counter) const {
```

Comment:
no need to use size_t && counter, because size_t is a basic data type like int, float.